-sandbox
# Behavioral Risk Factor Surveillance System (BRFSS)

- **Dataset**

BRFSS combined landline and cell phone age-adjusted prevalence data. The objective is to get an insight into the preventive health factors and risk behaviors linked to chronic diseases and to give information and disease prevalence that can affect healthy adults. The project aims to know the major health factors affecting and leading causes of death.

Dataset Source : https://healthdata.gov/dataset/Behavioral-Risk-Factor-Surveillance-System-BRFSS-A/m7ww-jcmt/data

d
- **Defining the location of the dataset**

In [0]:
filePath = "/FileStore/Behavioral_Risk_Factor_Surveillance_System__BRFSS__Prevalence_Data__2011_to_present_.csv"

rawDF = spark.read.csv(filePath, header="true", inferSchema="true", multiLine="true", escape='"')

display(rawDF)

Year Locationabbr Locationdesc Class Topic Question Response Break_Out Break_Out_Category Sample_Size Data_value Confidence_limit_Low Confidence_limit_High Display_order Data_value_unit Data_value_type Data_Value_Footnote_Symbol Data_Value_Footnote DataSource ClassId TopicId LocationID BreakoutID BreakOutCategoryID QuestionID ResponseID GeoLocation 2012 GU Guam Demographics Age What is your age? 55-64 years Overall Overall 296 13.1 11.4 14.9 1167 % Crude Prevalence null null BRFSS CLASS05 Topic02 66 BO1 CAT1 AGE RESP014 (13.444304, 144.793731) 2012 GU Guam Demographics Age What is your age? 35-44 years Female Gender 272 21.2 18.1 24.2 1174 % Crude Prevalence null null BRFSS CLASS05 Topic02 66 SEX2 CAT2 AGE RESP012 (13.444304, 144.793731) 2020 AL Alabama Chronic Health Indicators Depression Ever told you that you have a form of depression? Yes College graduate Education Attained 356 19.7 17.4 22.0 1 % Crude Prevalence null null BRFSS CLASS03 TOPIC17 1 EDUCA4 CAT5 ADDEPEV3 RESP046 (32.84057112200048, -86.63186076199969) 2020 AL Alabama Chronic Health Indicators Diabetes Have you ever been told by a doctor that you have diabetes? Yes, pregnancy-related White, non-Hispanic Race/Ethnicity 32 0.8 0.5 1.1 1 % Crude Prevalence null null BRFSS CLASS03 TOPIC18 1 RACE01 CAT4 DIABETE4 RESP051 (32.84057112200048, -86.63186076199969) 2020 AL Alabama Chronic Health Indicators Diabetes Have you ever been told by a doctor that you have diabetes? No, pre-diabetes or borderline diabetes Overall Overall 92 1.5 1.1 1.9 1 % Crude Prevalence null null BRFSS CLASS03 TOPIC18 1 BO1 CAT1 DIABETE4 RESP052 (32.84057112200048, -86.63186076199969) 2013 GU Guam Health Status Overall Health How is your general health? Excellent 18-24 Age Group 69 24.5 18.4 30.7 16 % Crude Prevalence null null BRFSS CLASS08 Topic41 66 AGE01 CAT3 GENHLTH RESP056 (13.444304, 144.793731) 2013 GU Guam Demographics Marital Status What is your marital status? Partnered H.S. or G.E.D. Education Attained 48 7.8 5.2 10.5 96 % Crude Prevalence null null BRFSS CLASS05 Topic38 66 EDUCA2 CAT5 MARITAL RESP038 (13.444304, 144.793731) 2013 PR Puerto Rico Demographics Age What is your age? 25-34 years Some post-H.S. Education Attained 183 20.4 17.4 23.3 98 % Crude Prevalence null null BRFSS CLASS05 Topic02 72 EDUCA3 CAT5 AGE RESP011 (18.220833, -66.590149) 2020 AL Alabama Colorectal Cancer Screening Blood Stool Test Adults aged 50-75 who have had a blood stool test within the past year (variable calculated from one or more BRFSS questions) Had a blood stool test in the past year Less than $15,000 Household Income 25 10.0 5.6 14.4 1 % Crude Prevalence null null BRFSS CLASS04 TOPIC08 1 INCOME1 CAT6 _RFBLDS4 RESP190 (32.84057112200048, -86.63186076199969) 2011 GU Guam Demographics Age What is your age? 45-54 years Hispanic Race/Ethnicity 29 18.5 11.1 25.8 70 % Crude Prevalence null null BRFSS CLASS05 Topic02 66 RACE3 CAT4 AGE RESP013 (13.444304, 144.793731) 2011 GU Guam Demographics Employment What is your employment status? Homemkr 65+ Age Group 19 11.0 5.3 16.8 69 % Crude Prevalence null null BRFSS CLASS05 Topic22 66 AGE09 CAT3 EMPLOY RESP029 (13.444304, 144.793731) 2019 KS Kansas Health Care Access/Coverage Last Checkup About how long has it been since you last visited a doctor for a routine checkup? Within the past year 45-54 Age Group 1235 76.5 74.1 78.9 1 % Crude Prevalence null null BRFSS CLASS07 TOPIC36 20 AGE04 CAT3 CHECKUP1 RESP063 (38.34774030000045, -98.20078122699965) 2020 AL Alabama Demographics Employment What is your employment status? No work >yr Overall Overall 92 2.0 1.5 2.6 1 % Crude Prevalence null null BRFSS CLASS05 TOPIC22 1 BO1 CAT1 EMPLOY1 RESP028 (32.84057112200048, -86.63186076199969) 2020 AL Alabama Demographics Employment What is your employment status? No work >yr Female Gender 52 1.7 1.2 2.2 1 % Crude Prevalence null null BRFSS CLASS05 TOPIC22 1 SEX2 CAT2 EMPLOY1 RESP028 (32.84057112200048, -86.63186076199969) 2013 GU Guam Demographics Age What is your age? 55-64 ye

In [0]:
rawDF.columns

Out[2]: ['Year',
 'Locationabbr',
 'Locationdesc',
 'Class',
 'Topic',
 'Question',
 'Response',
 'Break_Out',
 'Break_Out_Category',
 'Sample_Size',
 'Data_value',
 'Confidence_limit_Low',
 'Confidence_limit_High',
 'Display_order',
 'Data_value_unit',
 'Data_value_type',
 'Data_Value_Footnote_Symbol',
 'Data_Value_Footnote',
 'DataSource',
 'ClassId',
 'TopicId',
 'LocationID',
 'BreakoutID',
 'BreakOutCategoryID',
 'QuestionID',
 'ResponseID',
 'GeoLocation']

In [0]:
rawDF.printSchema()

root
 |-- Year: integer (nullable = true)
 |-- Locationabbr: string (nullable = true)
 |-- Locationdesc: string (nullable = true)
 |-- Class: string (nullable = true)
 |-- Topic: string (nullable = true)
 |-- Question: string (nullable = true)
 |-- Response: string (nullable = true)
 |-- Break_Out: string (nullable = true)
 |-- Break_Out_Category: string (nullable = true)
 |-- Sample_Size: integer (nullable = true)
 |-- Data_value: double (nullable = true)
 |-- Confidence_limit_Low: double (nullable = true)
 |-- Confidence_limit_High: double (nullable = true)
 |-- Display_order: integer (nullable = true)
 |-- Data_value_unit: string (nullable = true)
 |-- Data_value_type: string (nullable = true)
 |-- Data_Value_Footnote_Symbol: string (nullable = true)
 |-- Data_Value_Footnote: string (nullable = true)
 |-- DataSource: string (nullable = true)
 |-- ClassId: string (nullable = true)
 |-- TopicId: string (nullable = true)
 |-- LocationID: integer (nullable = true)
 |-- BreakoutID: strin

In [0]:
# create object and specify input and output column
OHE = OneHotEncoder(inputCols=['Locationdesc_Index', 'Class_Index','Topic_Index','Question_Index','Response_Index','Break_Out_Index','Break_Out_Category_Index'],outputCols=['Locationdesc_OHE', 'Class_OHE','Topic_OHE','Question_OHE','Response_OHE','Break_Out_OHE','Break_Out_Category_OHE'])

# transform the data
df = OHE.fit(df).transform(df)

# view and transform the data
df.select('Locationdesc', 'Locationdesc_Index', 'Locationdesc_OHE','Class','Class_Index','Class_OHE').show(10)

## Impute: Cast to Double

SparkML's `Imputer` requires all fields be of type double [Python](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html#pyspark.ml.feature.Imputer)/[Scala](https://spark.apache.org/docs/latest/api/scala/#org.apache.spark.ml.feature.Imputer). Let's cast all integer fields to double.

In [0]:
from pyspark.sql.functions import col
from pyspark.sql.types import IntegerType

integerColumns = [x.name for x in rawDF.schema.fields if x.dataType == IntegerType()]
doublesDF = rawDF

for c in integerColumns:
  doublesDF = doublesDF.withColumn(c, col(c).cast("double"))

columns = "\n - ".join(integerColumns)
print(f"Columns converted from Integer to Double:\n - {columns}")

Columns converted from Integer to Double:
 - Year
 - Sample_Size
 - Display_order
 - LocationID


Add in dummy variable if we will impute any value.

In [0]:
from pyspark.sql.functions import when

imputeCols = [
  "Year",
  "Sample_Size",
  "Display_order", 
  "LocationID"
]

for c in imputeCols:
  doublesDF = doublesDF.withColumn(c + "_na", when(col(c).isNull(), 1.0).otherwise(0.0))

In [0]:
from pyspark.ml.feature import Imputer

imputer = Imputer(strategy="median", inputCols=imputeCols, outputCols=imputeCols)

imputedDF = imputer.fit(doublesDF).transform(doublesDF)

- **Exploratory Data Analysis**

- **Year**

In [0]:
df1 = imputedDF.groupBy("Year").count()
display(df1)

Year count 2020.0 220975 2021.0 238463 2019.0 223058 2017.0 235303 2011.0 145364 2013.0 202751 2012.0 148824 2014.0 194798 2015.0 230101 2018.0 217579 2016.0 229714

In [0]:
df1.sort("Year")

+------+------+
|  Year| count|
+------+------+
|2011.0|145364|
|2012.0|148824|
|2013.0|202751|
|2014.0|194798|
|2015.0|230101|
|2016.0|229714|
|2017.0|235303|
|2018.0|217579|
|2019.0|223058|
|2020.0|220975|
|2021.0|238463|
+------+------+



In [0]:
display(df1.sort("Year"))

Year count 2011.0 145364 2012.0 148824 2013.0 202751 2014.0 194798 2015.0 230101 2016.0 229714 2017.0 235303 2018.0 217579 2019.0 223058 2020.0 220975 2021.0 238463

In [0]:
display(df1.sort("Year"))

Year count 2011.0 145364 2012.0 148824 2013.0 202751 2014.0 194798 2015.0 230101 2016.0 229714 2017.0 235303 2018.0 217579 2019.0 223058 2020.0 220975 2021.0 238463

Output can only be rendered in Databricks

The graph shows a steady increase in data from 2011 to 2021, with the highest amount of data collected being 238,463 in 2021.

-**Location**

In [0]:
df2 = imputedDF.groupBy("Locationdesc").count()
display(df2)

Locationdesc count Utah 43117 Hawaii 43111 Minnesota 43121 Ohio 43120 Arkansas 43050 Oregon 43087 Texas 43118 North Dakota 43025 Pennsylvania 43115 Connecticut 43043 Nebraska 43127 Vermont 43000 Nevada 43104 Puerto Rico 42619 Washington 43068 Illinois 43094 Oklahoma 43122 Virgin Islands 8748 District of Columbia 43064 Delaware 43091 Alaska 43094 New Mexico 43105 West Virginia 43076 Missouri 43108 Rhode Island 43037 Georgia 43112 Montana 43075 Virginia 43112 Michigan 43120 Guam 43025 North Carolina 43118 All States and DC (median) ** 1790 Wyoming 43080 Kansas 43123 New Jersey 38836 Maryland 43069 Alabama 43105 Arizona 43066 All States, DC and Territories (median) ** 1790 Iowa 43099 Massachusetts 43124 Kentucky 43115 Louisiana 43115 Mississippi 43069 New Hampshire 43024 Tennessee 43090 Florida 38626 Indiana 43111 Idaho 43048 South Carolina 43125 South Dakota 43091 California 43122 New York 43102 Wisconsin 43093 Colorado 43113 Maine 43078

In [0]:
display(df2.sort("count"))

Locationdesc count All States and DC (median) ** 1790 All States, DC and Territories (median) ** 1790 Virgin Islands 8748 Florida 38626 New Jersey 38836 Puerto Rico 42619 Vermont 43000 New Hampshire 43024 North Dakota 43025 Guam 43025 Rhode Island 43037 Connecticut 43043 Idaho 43048 Arkansas 43050 District of Columbia 43064 Arizona 43066 Washington 43068 Maryland 43069 Mississippi 43069 Montana 43075 West Virginia 43076 Maine 43078 Wyoming 43080 Oregon 43087 Tennessee 43090 Delaware 43091 South Dakota 43091 Wisconsin 43093 Illinois 43094 Alaska 43094 Iowa 43099 New York 43102 Nevada 43104 New Mexico 43105 Alabama 43105 Missouri 43108 Hawaii 43111 Indiana 43111 Georgia 43112 Virginia 43112 Colorado 43113 Pennsylvania 43115 Kentucky 43115 Louisiana 43115 Utah 43117 Texas 43118 North Carolina 43118 Ohio 43120 Michigan 43120 Minnesota 43121 Oklahoma 43122 California 43122 Kansas 43123 Massachusetts 43124 South Carolina 43125 Nebraska 43127

In [0]:
display(df2.sort("count"))

Locationdesc count All States and DC (median) ** 1790 All States, DC and Territories (median) ** 1790 Virgin Islands 8748 Florida 38626 New Jersey 38836 Puerto Rico 42619 Vermont 43000 New Hampshire 43024 North Dakota 43025 Guam 43025 Rhode Island 43037 Connecticut 43043 Idaho 43048 Arkansas 43050 District of Columbia 43064 Arizona 43066 Washington 43068 Maryland 43069 Mississippi 43069 Montana 43075 West Virginia 43076 Maine 43078 Wyoming 43080 Oregon 43087 Tennessee 43090 Delaware 43091 South Dakota 43091 Wisconsin 43093 Illinois 43094 Alaska 43094 Iowa 43099 New York 43102 Nevada 43104 New Mexico 43105 Alabama 43105 Missouri 43108 Hawaii 43111 Indiana 43111 Georgia 43112 Virginia 43112 Colorado 43113 Pennsylvania 43115 Kentucky 43115 Louisiana 43115 Utah 43117 Texas 43118 North Carolina 43118 Ohio 43120 Michigan 43120 Minnesota 43121 Oklahoma 43122 California 43122 Kansas 43123 Massachusetts 43124 South Carolina 43125 Nebraska 43127

Output can only be rendered in Databricks

-The majority of locations mentioned are in the midwest and south, with Nebraska (43127), South Carolina (43125), Kansas (43123), Oklahoma (43122) and Minnesota (43121). Massachusetts (43124) and California (43122) are also mentioned.

-**Class**

In [0]:
df3 = imputedDF.groupBy("Class").count()

In [0]:
display(df3.sort("count"))

Class count Prostate Cancer 11932 Hypertension Awareness 16084 HIV-AIDS 24415 Days of Poor Health 25320 E-Cigarette Use 25632 Fruits and Vegetables 27284 Injury 35479 Oral Health 37474 Women's Health 39818 Colorectal Cancer Screening 44687 Cholesterol Awareness 47854 Overweight and Obesity (BMI) 58538 Immunization 64091 Physical Activity 68632 Alcohol Consumption 87881 Health Status 102514 Tobacco Use 124422 Health Care Access/Coverage 179418 Chronic Health Indicators 457884 Demographics 807571

In [0]:
display(df3.sort("count"))

Class count Prostate Cancer 11932 Hypertension Awareness 16084 HIV-AIDS 24415 Days of Poor Health 25320 E-Cigarette Use 25632 Fruits and Vegetables 27284 Injury 35479 Oral Health 37474 Women's Health 39818 Colorectal Cancer Screening 44687 Cholesterol Awareness 47854 Overweight and Obesity (BMI) 58538 Immunization 64091 Physical Activity 68632 Alcohol Consumption 87881 Health Status 102514 Tobacco Use 124422 Health Care Access/Coverage 179418 Chronic Health Indicators 457884 Demographics 807571

Output can only be rendered in Databricks

The majority of questions asked in the BRFSS survey are related to demographics, such as age, gender, ethnicity, education level, and employment status. Other topics covered by the survey include health-related behaviors, chronic health conditions, health insurance coverage, access to health care, and use of preventive services. The survey also collects data on physical activity, alcohol and tobacco use, nutrition, mental health, and sexual behavior.

- **Topic**

In [0]:
df4 = imputedDF.groupBy("Topic").count()

In [0]:
display(df4.sort("count"))

Topic count Shingle Vaccination 4559 Colonoscopy 4930 Vision 7086 Sigmoidoscopy 9174 Tetanus Shot 10300 USPSTF Recommendations 10714 All Teeth Removed 11049 PSA Test 11932 Aerobic Activity 13112 Strength Activity 13112 Physical Activity Index 13112 Drink and Drive 13137 Teeth Removed 13212 Dental Visit 13213 Internet 13215 Fruit Consumption 13642 Vegetable Consumption 13642 Pap Test 14497 High Blood Pressure 16084 Cholesterol High 16084 Hearing 16772 Blood Stool Test 19869 Seatbelt Use 22342 Health Care Cost 24413 Veteran Status 24414 HIV Test 24415 Pneumonia Vaccination 24612 Flu Shot 24620 Healthy Days 25320 Mammogram 25321 E-Cigarette Use 25632 Sex 27176 Under 65 Coverage 28127 Skin Cancer 29230 Kidney 29269 Other Cancer 29287 COPD 29287 Heavy Drinking 29290 Health Care Coverage 29293 Depression 29295 Current Smoker Status 29295 Binge Drinking 29295 Exercise 29296 Alcohol Consumption 29296 Fair or Poor Health 29297 Cholesterol Checked 31770 Smokeless Tobacco 36552 Rent/Own Home 36619 Personal Care Provider 36620 Education 49259 Diabetes 58168 BMI Categories 58538 Smoker Status 58575 Asthma 58587 Income 60357 Last Checkup 60965 Age 66881 Overall Health 73217 Race 73437 Arthritis 83175 Number of Children 87646 Marital Status 87757 Cardiovascular Disease 104500 Employment 116940 Disability status 147098

In [0]:
display(df4.sort("count"))

Topic count Shingle Vaccination 4559 Colonoscopy 4930 Vision 7086 Sigmoidoscopy 9174 Tetanus Shot 10300 USPSTF Recommendations 10714 All Teeth Removed 11049 PSA Test 11932 Aerobic Activity 13112 Strength Activity 13112 Physical Activity Index 13112 Drink and Drive 13137 Teeth Removed 13212 Dental Visit 13213 Internet 13215 Fruit Consumption 13642 Vegetable Consumption 13642 Pap Test 14497 High Blood Pressure 16084 Cholesterol High 16084 Hearing 16772 Blood Stool Test 19869 Seatbelt Use 22342 Health Care Cost 24413 Veteran Status 24414 HIV Test 24415 Pneumonia Vaccination 24612 Flu Shot 24620 Healthy Days 25320 Mammogram 25321 E-Cigarette Use 25632 Sex 27176 Under 65 Coverage 28127 Skin Cancer 29230 Kidney 29269 Other Cancer 29287 COPD 29287 Heavy Drinking 29290 Health Care Coverage 29293 Depression 29295 Current Smoker Status 29295 Binge Drinking 29295 Exercise 29296 Alcohol Consumption 29296 Fair or Poor Health 29297 Cholesterol Checked 31770 Smokeless Tobacco 36552 Rent/Own Home 36619 Personal Care Provider 36620 Education 49259 Diabetes 58168 BMI Categories 58538 Smoker Status 58575 Asthma 58587 Income 60357 Last Checkup 60965 Age 66881 Overall Health 73217 Race 73437 Arthritis 83175 Number of Children 87646 Marital Status 87757 Cardiovascular Disease 104500 Employment 116940 Disability status 147098

Output can only be rendered in Databricks

The graph shows that disability status is the most reported chronic health condition among adults, followed by cardiovascular disease, arthritis, and employment.